### ISSUE TITLE
Check if slot is filled and control the flow in stories

### ISSUE THREAD

**Issue by saivisal1** [RasaHQ/rasa_core#637]
**Rasa Core version**: 0.9.6

**Python version**: 3.7

**Operating system** (windows, osx, ...): OS X

**Issue**: 

I have a slot city which is used for 2 custom actions weather check and petrol price check. If slot is not set before the custom action call for weather or price check it should do action utter_ask_city. if slot is filled with city name it should do the action requested. How can I handle this situation?

Can i call another custom action from a custom action?
or can have a checkpoint that checks if the slot is filled?

**Content of domain file** (if used & relevant):
```
slots:
    fuel:
        type: text
    city:
        type: text

    
intents:
    - greet
    - confirm
    - deny
    - goodbye
    - price_check
    - thankyou
    - restaurant_search
    - video_search
    - weather_check

entities:
    - fuel
    - city

templates:
    utter_greet:
        - 'Hello! How can I help?'
        - 'Hi! How can I help?'
        - 'Welcome, how may I assist you?'
    utter_goodbye:
        - 'Talk to you later.'
        - 'Bye bye'
    utter_thankyou:
        - 'Cool!!'
        - 'You are welcome!'
        - 'Its my pleasure'
    utter_deny:
        - 'Sorry!!'
    utter_affirm:
        - 'Ok'
        - 'Yes'
    utter_default:
        - 'I am not very sure what you said.'
        - "Maybe I didn't understand you properly. Could you please reword your question or type your complete query?"
        - "I didn't get what you said. Can you please rephrase the query?"
        - "A little lost. Apologies. Could you rephrase?"
        - "I've dropped the ball. Could you rephrase?"
        - "Oops. Signal fuzzy. Could you rephrase?"
        - "That request lies outside my powers, although I'm always learning. I can tell you about the fuel prices, suggest good restarants, updates on weather updates, and answer questions. And I'm always learning."
    
actions:
    - utter_greet
    - utter_goodbye
    - utter_thankyou
    - utter_deny
    - utter_default
    - utter_affirm
    - actions.AllSlotsReset
    - actions.ActionPriceCheck
    - actions.ActionAskCity
    - actions.ActionAskFuel
    - actions.ActionRestaurantSearch
    - actions.ActionVideoSearch
    - actions.ActionWeatherCheck
```

**Comment by tmbo** [comment]
Can you please add the stories you wrote as well?

**Comment by saivisal1** [comment]
@tmbo 

Here is the stories which I wrote.

`## Story 1
* greet
    - utter_greet

## Story 2
* price_check{"city": "delhi","fuel": "petrol"}
    - slot{"city": "delhi"}
    - slot{"fuel": "petrol"}
    - price_api
    
## Story 3
* price_check{"city": "delhi"}
    - slot{"city": "delhi"}
    - utter_ask_fuel
* price_check{"fuel": "petrol"}
    - slot{"fuel": "petrol"}
    - price_api
    
## Story 4
* price_check{"fuel": "petrol"}
    - slot{"fuel": "petrol"}
    - action_ask_city
* price_check{"city": "delhi"}
    - slot{"city": "delhi"}
    - price_api

## Story 4.1
* price_check{"city": "delhi"}
    - slot{"city": "delhi"}
    - price_api

## Story 5
* thankyou
    - utter_thankyou

## Story 6
* goodbye
    - utter_goodbye
    - action_slot_reset

## Story 7
* restaurant_search
    - action_restaurant

## Story 8
* video_search
    - action_video

## Story 9
* weather_check{"city": "Bengaluru"}
    - slot{"city": "Bengaluru"}
    - action_weather

## Story 10
* weather_check
    - action_ask_city
* weather_check{"city": "Chennai"}
    - slot{"city": "Chennai"}
    - action_weather`

**Comment by amn41** [comment]
I just tried this out on your data and it seems to already do what you want (I replaced the custom actions with simple utterances)

I ran 
`python -m rasa_core.train -d domain.yml -s stories.md -o models/dialogue`
then
` python -m rasa_core.run -d models/dialogue`

and got
```
Bot loaded. Type a message and press enter: 
/greet
Welcome, how may I assist you?
/price_check
what city?
/price_check{"city": "delhi"}
price is x
```

**Comment by saivisal1** [comment]
Try this:

User: Hi
Bot: Hi how may i assist you?
User: what is the weather in Delhi
Bot: its X degrees
User: what is the price?
Bot: its X price
User: thank you

User: Hi
Bot: Hi how may i assist you?
User: what is the weather ?
Bot: which city?
User: CityY
Bot: its X degrees
User: what is the fuel price?
Bot: its X price
User: thank you

User: Hi
Bot: Hi how may i assist you?
User: what is the fuel price ?
Bot: which city?
User: CityY
Bot: its X degrees

I wanted to ask for slot if its not filled. Same slot am using for 2 different actions. If slot is already filled the action requested should happen. If slot is filled in any of the previous stories with the user again the utter_ask_slot should not happen.

How can i achieve this?

**Comment by saivisal1** [comment]
Hi @amn41 ,

I am going to add new 10 more actions that use the same slot. So I need to make stories in all permutations and combination. 

Is there any other solutions?

**Comment by amn41** [comment]
I would strongly recommend giving your bot to people to test before adding a further 10 actions. It's much easier (and more useful) to collect stories from people actually talking to your bot than trying to write them by hand!

**Comment by amn41** [comment]
alternatively you can define these as `FormActions` with location as a required slot, see http://core.rasa.com/patterns.html

**Comment by saivisal1** [comment]
Ok i tried FormActions it was not working. The issue was I as using `run()` instead of  `submit()` in the custom action. 

Thanks @tmbo and @amn41
